dmutils/request_id.py

```diff
@@ -1,50 +1,90 @@
+
+    @property
+    def trace_id(self):
+        if not hasattr(self, "_trace_id"):
+            self._trace_id = self._get_first_header(current_app.config['DM_TRACE_ID_HEADERS']) or str(uuid.uuid4())
+        return self._trace_id
+
+    @property
+    def span_id(self):
+        if not hasattr(self, "_span_id"):
+            self._span_id = self._get_first_header(current_app.config['DM_SPAN_ID_HEADERS'])
+        return self._span_id
+
+    @property
+    def parent_span_id(self):
+        if not hasattr(self, "_parent_span_id"):
+            self._parent_span_id = self._get_first_header(current_app.config['DM_PARENT_SPAN_ID_HEADERS'])
+        return self._parent_span_id
+
+    def _get_first_header(self, header_names):
+        """
+        Returns value of request's first present (and Truthy) header from header_names
+        """
+        for header_name in header_names:
+            if header_name in self.headers and self.headers[header_name]:
+                return self.headers[header_name]
         else:
-            return str(uuid.uuid4())
+            return None
+
+    def get_onwards_request_headers(self):
+        """
+            Headers to add to any further (internal) http api requests we perform if we want that request to be
+            considered part of this "trace id"
+        """
+        return dict(chain(
+            (
+                (header_name, self.trace_id)
+                for header_name in current_app.config['DM_TRACE_ID_HEADERS']
+                if self.trace_id
+            ),
+            (
+                (header_name, self.span_id)
+                for header_name in current_app.config['DM_SPAN_ID_HEADERS']
+                if self.span_id
+            ),
+        ))
 
 
 class ResponseHeaderMiddleware(object):
-    def __init__(self, app, request_id_header):
+    def __init__(self, app, trace_id_headers):
         self.app = app
-        self.request_id_header = request_id_header
+        self.trace_id_headers = trace_id_headers
 
     def __call__(self, environ, start_response):
         def rewrite_response_headers(status, headers, exc_info=None):
-            if self.request_id_header not in dict(headers).keys():
-                headers = headers + [(
-                    self.request_id_header,
-                    str(request.request_id)
-                )]
+            headers = headers + [
+                (header_name, str(request.request_id),)
+                for header_name in self.trace_id_headers
+                if header_name not in dict(headers)
+            ]
 
             return start_response(status, headers, exc_info)
 
         return self.app(environ, rewrite_response_headers)
 
 
 def init_app(app):
-    app.config.setdefault('DM_REQUEST_ID_HEADER', 'DM-Request-ID')
-    app.config.setdefault('DM_DOWNSTREAM_REQUEST_ID_HEADER', '')
+    app.config.setdefault("DM_TRACE_ID_HEADERS", (
+        (app.config.get("DM_REQUEST_ID_HEADER") or "DM-Request-ID"),
```

Comment:
Are `DM_REQUEST_ID_HEADER` and `DM_DOWNSTREAM_REQUEST_ID_HEADER` defined anywhere as to what their intention is?

I'm quite confused about what the later is actually for...

Reply:
Yup this is it exactly. What it turns out to mean is "a request header we are receiving as the *downstream* of a master proxy. It was originally there because I think our **original** hosting setup used to provide us with an id header similar to zipkin. But I found this all very confusing and thought it would be much simpler and clearer to just implement the ability to look for multiple header names in priority order in a much more general, less opinionated & situation-specific way.

Really I don't want to document their intentions. I want them to die.

Reply:
Yeah getting rid of them sounds good